src/dawn/SIR/ASTOps.cpp

```diff
@@ -0,0 +1,93 @@
+//===--------------------------------------------------------------------------------*- C++ -*-===//
+//                          _
+//                         | |
+//                       __| | __ ___      ___ ___
+//                      / _` |/ _` \ \ /\ / / '_  |
+//                     | (_| | (_| |\ V  V /| | | |
+//                      \__,_|\__,_| \_/\_/ |_| |_| - Compiler Toolchain
+//
+//
+//  This file is distributed under the MIT License (MIT).
+//  See LICENSE.txt for details.
+//
+//===------------------------------------------------------------------------------------------===//
+
+#include "ASTOps.h"
+
+namespace dawn {
+
+namespace {
+template <typename T>
+typename std::vector<T>::const_iterator
+arrayRefIteratorToVectorIterator(const std::vector<T>& vector,
+                                 typename ArrayRef<T>::iterator arrayRefIterator,
+                                 typename ArrayRef<T>::iterator beginArrayRefIterator) {
+  typename std::vector<T>::const_iterator vecIt = vector.begin();
+  return std::next(vecIt, arrayRefIterator - beginArrayRefIterator);
+}
+
+template <typename T>
+typename ArrayRef<T>::iterator
+vectorIteratorToArrayRefIterator(const std::vector<T>& vector,
+                                 typename std::vector<T>::const_iterator vectorIterator) {
+  return vector.empty() ? nullptr : vector.data() + (vectorIterator - vector.begin());
+}
+
+} // namespace
+
+// Input contract: iterator root must be a BlockStmt
+template <>
+ASTNodeIterator<Stmt, true> ASTOps::insertAfter(std::shared_ptr<Stmt> stmt,
```

Comment:
(see also https://github.com/Stagno/dawn/pull/2#discussion_r306352120)
Agree that algorithms should not reside inside iterators.
Problem is now aggravated since `ASTOps` needs access to private/protected members/methods of `ASTNodeIterator`, `ASTNodeIteratorImpl`, `SubtreeIteratorImpl`, `VoidIteratorImpl`.
I guess the question is: is there a way to let a class have a "public" set of APIs and a set of APIs reserved for algorithms that need to go "deeper"?